src/notebook.rs

```diff
@@ -0,0 +1,432 @@
+    notebook_selector: Vec<NotebookSelector>,
+    /// Whether save notification should be forwarded to
+    /// the server. Will only be honored if mode === `notebook`.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    save: Option<bool>,
+    /// The id used to register the request. The id can be used to deregister
+    /// the request again. See also Registration#id.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    id: Option<String>,
+}
+
+/// A notebook cell text document filter denotes a cell text
+/// document by different properties.
+///
+/// @since 3.17.0
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NotebookCellTextDocumentFilter {
+    /// A filter that matches against the notebook
+    /// containing the notebook cell. If a string
+    /// value is provided it matches against the
+    /// notebook type. '*' matches every notebook.
+    notebook: Notebook,
+    /// A language id like `python`.
+    ///
+    /// Will be matched against the language id of the
+    /// notebook cell document. '*' matches every language.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    language: Option<String>,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum NotebookSelector {
+    ByNotebook(NotebookSelectorByNotebook),
+    ByCells(NotebookSelectorByCells),
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NotebookSelectorByNotebook {
+    /// The notebook to be synced. If a string
+    /// value is provided it matches against the
+    /// notebook type. '*' matches every notebook.
+    notebook: Notebook,
+    /// The cells of the matching notebook to be synced.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    cells: Option<Vec<NotebookCellSelector>>,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NotebookSelectorByCells {
+    /// The notebook to be synced. If a string
+    /// value is provided it matches against the
+    /// notebook type. '*' matches every notebook.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    notebook: Option<Notebook>,
+    /// The cells of the matching notebook to be synced.
+    cells: Vec<NotebookCellSelector>,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NotebookCellSelector {
+    language: String,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum Notebook {
+    String(String),
+    NotebookDocumentFilter(NotebookDocumentFilter),
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum NotebookDocumentFilter {
+    ByType(NotebookDocumentFilterByType),
+    ByScheme(NotebookDocumentFilterByScheme),
+    ByPattern(NotebookDocumentFilterByPattern),
+}
```

Comment:
Mm, I see where you're coming from, but I'm not entirely convinced that this should simply use optional fields. You might be correct in that optional fields might be more usable than the other, but more so for those well versed with the spec.

I still think one should try to leverage the type system as much as possible when it comes to defining the domain.  The current solution is one way of doing so. Silent errors and unpleasant surprises are pretty easy to introduce if not. 

I'm nonetheless open to changing this as suggested, but I think I would like to receive some input from others before deciding on anything. (If that's even up to me.) 😊

Reply:
I agree with you, but think that the interface of NotebookDocumentFilter should be consistent with DocumentFilter, because they are basically the same thing. So we should either change this PR or change the interface of DocumentFilter.

Maybe @Marwes or @ebkalderon can help us decide.

Reply:
I feel like the patch as is captures the specification better than having 3 optional fields.
The only thing I would consider is since the spec doesn't define the `NotebookDocumentFilterBy*` types,
you could consider inlining those structures into the enum variants themselves like:

```
enum NotebookDocumentFilter {
  ByType{notebook_type: String, scheme: Option<String>, pattern: Option<String>},
  ...
}
```

I feel like the `DocumentFilter` case is different, since there the spec is actually defining it as a single struct with 3 optional fields, but `NetworkDocumentFilter` defines it differently, as one of any 3 structs each with 1 required field, and 2 optional fields... `enum` captures that latter definition better than struct.

I'm overall not very opinionated on whether its worthwhile to eliminate these intermediate types, 
and while I do use the library, I haven't really contributed to this library before so my opinion shouldn't really hold much weight honestly.

Reply:
I can totally get behind inlining the variants! Great idea :)

Reply:
I like how https://github.com/gluon-lang/lsp-types/pull/268/commits/df3eeea62503aae0405fe52a99012a23809dc8ca captures things, as I also agree that the spec itself differs from how it presents `NotebookDocumentFilter` and `DocumentFilter` 🤗

Reply:
Inlining it seems like the right trade-off for now. Ideally the LSP spec should homogenize how it defines these two types though (and we should as well), but for now having them be different seems fine, we might be able to find out which is preferred by hearing from people who use both.